app/src/main/java/ca/unb/mobiledev/reflexrevolution/instructions/TypeInstruction.java

```diff
@@ -0,0 +1,128 @@
+        setUpRandomWordList();
+    }
+
+    //Adapted from https://stackoverflow.com/a/48865091
+    private void setUpRandomWordList(){
+        try {
+            File outputDir = context.getCacheDir();
+            File destFile = new File(outputDir, "TypeWords.txt");
+            destFile.delete();
+            OutputStream dest = new FileOutputStream(destFile, true);
+            InputStream src = context.getAssets().open("TypeWords.txt", AssetManager.ACCESS_STREAMING);
+            byte[] buff = new byte[100 * 1024];
+            for (; ; ) {
+                int cnt = src.read(buff);
+                if (cnt <= 0)
+                    break;
+                dest.write(buff, 0, cnt);
+            }
+            dest.flush();
+            dest.close();
+            wordListFile = new RandomAccessFile(destFile, "r");
+
+            //Get number of lines in file
+            fileLength = wordListFile.length();
+        } catch(IOException e){ e.printStackTrace(); }
+    }
+
+    @Override
+    public void init() {
+        super.init();
+        //Get random word to type
+        try {
+            //Subtracting file line size outside random so there's an equal chance of getting the first string in the list as the others
+            int pos = (rand.nextInt((int)fileLength) - SIZE_OF_FILE_LINE);
+            if(pos < 0) pos = 0;
+            wordListFile.seek(pos);
+            if(pos != 0) wordListFile.readLine();
+            word = wordListFile.readLine();
+        } catch (IOException e) { e.printStackTrace(); }
+    }
+
+    @Override
+    public void display() {
+        addTextView(word, SMALL_TEXT_SIZE);
+        addTextView("TYPE", DEFAULT_TEXT_SIZE);
+        field = new EditText(context);
+
+        //Listener that checks for text being updated
+        field.addTextChangedListener(new TextWatcher() {
+            //If text equals current word, trigger success
+            @Override
+            public void onTextChanged(CharSequence s, int start, int before, int count) {
+                if(s.toString().toUpperCase().equals(word)){
+                    //Hide keyboard
+                    disable();
+                    success();
+                }
+            }
+
+            @Override
+            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+                //Auto-generated
+            }
+            @Override
+            public void afterTextChanged(Editable s) {
+                //Auto-generated
+            }
+        });
+        addView(field);
+        enable();
+    }
+
+    //Select field and force the keyboard to show
+    @Override
+    public void enable() {
+        field.requestFocus();
+        keyboardDisplayManager.toggleSoftInput(InputMethodManager.SHOW_FORCED, InputMethodManager.HIDE_IMPLICIT_ONLY);
+    }
+
+    //Close keyboard
+    @Override
+    public void disable() {
+        keyboardDisplayManager.hideSoftInputFromWindow(field.getWindowToken(), 0);
+    }
```

Comment:
Why are we using the flag value of 0? According to the documentation, 0 corresponds to the [`RESULT_UNCHANGED_SHOWN`](https://developer.android.com/reference/android/view/inputmethod/InputMethodManager#RESULT_UNCHANGED_SHOWN) flag which doesn't make sense in this case. Looking at the documentation, it looks like `HIDE_IMPLICIT_ONLY` would make more sense.
```java
keyboardDisplayManager.hideSoftInputFromWindow(field.getWindowToken(), InputMethodManager.HIDE_IMPLICIT_ONLY);
```

P.S.: Don't forget to do the same for the Dial instruction.

Reply:
Yeah, I think I just saw an example on a forum that used 0, but after looking at the documentation, HIDE_IMPLICIT_ONLY makes more sense.